lib/tasks/find_records_missing.rake

```diff
@@ -0,0 +1,158 @@
+  end
+end
+
+desc "Finds database records that are missing parent records"
+
+task "find_records_missing_parents", [:max_per_model] => :environment do |t, args|
+  ActiveRecord::Base.logger.level = :info
+  args.with_defaults(:max_per_model => 1000)
+  must_have_parents.each do |model_name, assocs| 
+    puts "\nFinding #{model_name} records that are missing a parent record by belongs_to assocs #{assocs} ..."
+    model = all_models[model_name.to_s]
+    missing = missing_parents(model, assocs, args.max_per_model)
+    if missing.length > 0
+      puts "Model #{model.name} has missing parents in records: #{JSON.pretty_generate(missing)}"
+    end
+    puts get_total_message(args.max_per_model, model)
+  end
+end
+
+desc "Finds database records that appear to be test data"
+
+task "find_test_records", [:max_per_model] => :environment do |t, args|
+  ActiveRecord::Base.logger.level = :info
+  args.with_defaults(:max_per_model => 1000)
+  all_models.select.each do |name, model| 
+    puts "\nFinding #{name} records that contain 'test' or 'demo' in their name ..."
+    model.order(created_at: :desc).limit(args.max_per_model).map do |record|
+      if is_test_record(record)
+        owner = record.respond_to?(:user) && record.user ? record.user.email : "unknown"
+        puts "Record ##{record.id} '#{record.name}' owned by #{owner} appears to be test data"
+      end
+    end
+    puts get_total_message(args.max_per_model, model)
+  end
+end
+
+private 
+
+def is_test_record(record)
+  record.respond_to?(:name) && record.name && 
+    record.name.match(/\b(test|demo|benchmark|asdf)\b/)
+end
+
+def missing_children(model, assocs, max_per_model)
+  format_missing(model.order(created_at: :desc).limit(max_per_model).map do |record|
+    assocs
+      .map { |assoc| [record.id, assoc, record.send(assoc).pluck(:id)] }
+      .select { |data| data.last.empty? }
+      .map { |data| [data[0], data[1]] }
+  end)
+end
+
+def missing_parents(model, assocs, max_per_model)
+  format_missing(model.order(created_at: :desc).limit(max_per_model).map do |record|
+    assocs
+      .map { |assoc| [record.id, assoc, record.send(assoc)] }
+      .select { |data| data.last.nil? }
+      .map { |data| [data[0], data[1]] }
+  end)
+end
+
+def format_missing(missing)
+  missing
+  .select {|m| m.length > 0}
+  .flatten(1)
+  .group_by{|m| m[0]}
+  .map { |k, v| [k, v.map { |e| e[1] }] }
+  .to_h
+end
+
+# Run this in the rails console
+def get_initial(assoc_type=:has_many)
+  models = ApplicationRecord.descendants
+  models.map {|m| [m.name, m.reflect_on_all_associations(assoc_type).map {|a| a.name}]}
+    .select { |p| p[1].length >0 }
+    .to_h
+end
+
+# Data first obtained by get_initial(:has_many)
+def must_have_children
+  {
+    :AlignmentConfig => [:pipeline_runs],
+    :HostGenome => [:samples],
+    :MetadataField => [:metadata],
+    :PipelineRun => [
+       :pipeline_run_stages,
+       :output_states,
+       # :taxon_counts,
+       :job_stats,
```

Comment:
PipelineRun: I think job_stats are not required either (if a run fails really early). See pipeline_run#monitor_results (compiles the stats file after starting to check output states)